### ISSUE TITLE
"unexpected eof while reading" is back in OpenSSL 3 with FTP

### ISSUE THREAD

**Issue by Zenju** [curl/curl#7800]
The "unexpected eof while reading" is back in OpenSSL 3, as expected. 
https://github.com/curl/curl/issues/5138

The unexpected EOF seems to only affect some (not well-behaved?) servers. I don't have a test server myself, but one of my software's users has: https://freefilesync.org/forum/viewtopic.php?t=8800

OpenSSL 3 added [ SSL_OP_IGNORE_UNEXPECTED_EOF](https://www.openssl.org/docs/man3.0/man3/SSL_CTX_set_options.html#SSL_OP_IGNORE_UNEXPECTED_EOF)
Maybe this should be added to [ctx_options](https://github.com/curl/curl/blob/2b7e56aab353188e7919cd941150abb77ffc4d97/lib/vtls/openssl.c#L2726) to continue with the old EOF handling?
I made a quick test (=hack) without much understanding of the consequences in libcurl. It seems, just adding SSL_OP_IGNORE_UNEXPECTED_EOF creates other with libcurl (=> empty directory listings)
https://freefilesync.org/forum/viewtopic.php?t=8800#p31369

So I'm not sure what the best fix for curl is. Maybe it's to add OpenSSL version-dependent  code like:
````````
#if OPENSSL_VERSION_NUMBER >= 0x30000000L /*OpenSSL 3.0.0*/ || \
    OPENSSL_VERSION_NUMBER == 0x1010105fL /*OpenSSL 1.1.1e*/
            const auto ec = ::ERR_peek_last_error();
            if (sslError == SSL_ERROR_SSL && ERR_GET_REASON(ec) == SSL_R_UNEXPECTED_EOF_WHILE_READING) //EOF: only expected for HTTP/1.0
#else //obsolete handling: https://github.com/openssl/openssl/issues/10880#issuecomment-575746226
            if ((sslError == SSL_ERROR_SYSCALL && ::ERR_peek_last_error() == 0)) //EOF: only expected for HTTP/1.0
#endif
````````

**Comment by bagder** [comment]
Yes, "the issue is back", but isn't this a server-side problem? Do we really serve our users best by just silently continuing to pretend this error didn't occur? Maybe this is rather the time for us to step up our game and point fingers instead?

**Comment by bagder** [comment]
< I'm wondering if libssh2 is also affected by the change in behavior with OpenSSL 3

libssh2 doesn't use OpenSSL for TLS though so I don't think this issue applies there.

**Comment by bagder** [comment]
If we don't think its a bug, it's more of a "documented behavior" and not a  known bug... but I don't know where the best place would be for adding such a documentation note!

Closing this issue. Not a bug.

**Comment by codgunt** [comment]
This is not curl's fault of course but not providing a workaround means that anyone relying on libcurl in PHP to work with webservices behind upstream IIS installs is left without recourse.   This issue is present as of today in PHP 7.4.30 and PHP 8.1.6 at least with the Fedora 36 distro.

**Comment by kroeckx** [comment]
OpenSSL can not tell that all data was received or not if the close_notify is not received. It looks like a truncation attack to OpenSSL, so it's reporting that error. But for many protocols there can be ways to see if all data was received or not. I don't know enough about all HTTP versions and options, but as far as I know, it should be clear in most cases how much data you are going to receive. If it's clear how much data you're going to receive, and you check that you did receive all that data, setting that option is safe.

**Comment by jay** [comment]
Yes, usually there is a known termination point for fetching a body (ie server says content length or chunked encoding 0 or file size etc) and curl won't read past that however if there is no known termination point in that case curl will keep reading.

**Comment by fifofonix** [comment]
I'm getting this same error: `curl: (35) error:0A000126:SSL routines::unexpected eof while reading`, but the behaviour is confusing to me and perhaps someone can shed some light?

Basically, I have some CICD docker-in-docker multi-architecture jobs running using qemu and on x86 fedora36 a curl command succeeds, yet the exact same curl command fails on aarch64 fedora36.

The curl version and all sub-components in both archs is the same:
```
curl 7.82.0 (aarch64-redhat-linux-gnu) libcurl/7.82.0 OpenSSL/3.0.2 zlib/1.2.11 brotli/1.0.9 libidn2/2.3.2 libpsl/0.21.1 (+libidn2/2.3.2) libssh/0.9.6/openssl/zlib nghttp2/1.46.0 OpenLDAP/2.6.1
Release-Date: 2022-03-05
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets
```

The URL causing the issue is a golang download: https://storage.googleapis.com/golang/go1.17.1.linux-arm64.tar.gz
(Obviously in my docker jobs this URL is going to be different for each arch but for reproduction purposes I have just used this one URL).

**Comment by jay** [comment]
> The URL causing the issue is a golang download: https://storage.googleapis.com/golang/go1.17.1.linux-arm64.tar.gz

Use verbose -v. The server should send a content length and curl should download just that amount of data.
~~~
< content-length: 102606384
~~~

You should then immediately see in the progress meter that it is going to download 97.8 megabytes total
~~~
 19 97.8M   19 18.9M    0     0  11.8M      0  0:00:08  0:00:01  0:00:07 11.8M
~~~

If someone else with arm could try to reproduce this that would be helpful, I tried with x64 and can't reproduce.

~~~
curl 7.84.0-DEV (x86_64-pc-linux-gnu) libcurl/7.84.0-DEV OpenSSL/3.0.4 zlib/1.2.8 nghttp2/1.48.0 librtmp/2.3 OpenLDAP/2.4.42
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS Debug HSTS HTTP2 HTTPS-proxy IPv6 Largefile libz NTLM NTLM_WB SSL threadsafe TLS-SRP TrackMemory UnixSockets
~~~~

**Comment by fifofonix** [comment]
Not reproducible on an arm vm running fedora36.  Could it be possible that this is somehow due to qemu?  The scenario where I have been getting this issue is when doing a an arm architecture build on an x86 vm.

**Comment by jay** [comment]
I don't know. The verbose output should provide some clue. If you have a way to reproduce and think it's a bug in curl then please open a new issue.

**Comment by oranges13** [comment]
> I don't think we should ignore EOF by default in OpenSSL 3. We gave this plenty of time and so did OpenSSL. The user could set the flag via [CURLOPT_SSL_CTX_FUNCTION](https://curl.se/libcurl/c/CURLOPT_SSL_CTX_FUNCTION.html).

Can you clarify how to set this flag and what value to use? I believe we are running up against this issue currently.

**Comment by jay** [comment]
> Can you clarify how to set this flag and what value to use? I believe we are running up against this issue currently.

[CURLOPT_SSL_CTX_FUNCTION](https://curl.se/libcurl/c/CURLOPT_SSL_CTX_FUNCTION.html)
[SSL_OP_IGNORE_UNEXPECTED_EOF](https://www.openssl.org/docs/man3.0/man3/SSL_CTX_set_options.html#SSL_OP_IGNORE_UNEXPECTED_EOF)

Here is an example
~~~c
#include <stdio.h>
#include <stdlib.h>
#include <curl/curl.h>
#include <openssl/ssl.h>

CURLcode ssl_ctx_callback(CURL *curl, void *ssl_ctx, void *userptr)
{
  SSL_CTX_set_options((SSL_CTX *)ssl_ctx, SSL_OP_IGNORE_UNEXPECTED_EOF);
  return CURLE_OK;
}

int main(int argc, char *argv[])
{
  CURL *curl;

  /* The ssl_ctx_callback must be matched to the SSL library used by libcurl.
     Since our ssl_ctx_callback is using OpenSSL then so must libcurl.
     To check this we try explicitly setting the backend to OpenSSL.

     curl_global_sslset must be called before any other libcurl function and
     will fail if libcurl was not built with the selected backend. */
  if(curl_global_sslset(CURLSSLBACKEND_OPENSSL, NULL, NULL)
     != CURLSSLSET_OK) {
    fprintf(stderr, "Fatal: Failed setting CURLSSLBACKEND_OPENSSL.\n");
    return EXIT_FAILURE;
  }

  /* curl_global_init should be called before any other libcurl function,
     with the exception of curl_global_sslset. */
  if(curl_global_init(CURL_GLOBAL_ALL)) {
    fprintf(stderr, "Fatal: Failed initializing libcurl.\n");
    return EXIT_FAILURE;
  }

  curl = curl_easy_init();
  curl_easy_setopt(curl, CURLOPT_SSL_CTX_FUNCTION, ssl_ctx_callback);
  curl_easy_setopt(curl, CURLOPT_URL, "https://brokenwebsite");
  curl_easy_perform(curl);

  return EXIT_SUCCESS;
}
~~~

If you aren't in control of libcurl initialization you should still check that the backend is OpenSSL before setting a CTX callback that expects an OpenSSL object. For example:
~~~c
struct curl_tlssessioninfo *tls_backend_info = NULL;

if(curl_easy_getinfo(curl, CURLINFO_TLS_SSL_PTR,
                     &tls_backend_info) != CURLE_OK ||
   tls_backend_info->backend != CURLSSLBACKEND_OPENSSL)
{
  // abort or do not set the ctx function
}
~~~

**Comment by oranges13** [comment]
> > Can you clarify how to set this flag and what value to use? I believe we are running up against this issue currently.
> 
> [CURLOPT_SSL_CTX_FUNCTION](https://curl.se/libcurl/c/CURLOPT_SSL_CTX_FUNCTION.html) [SSL_OP_IGNORE_UNEXPECTED_EOF](https://www.openssl.org/docs/man3.0/man3/SSL_CTX_set_options.html#SSL_OP_IGNORE_UNEXPECTED_EOF)
> 
> Here is an example

We are running into this issue with PHP specifically -- this looks like a different language. Any tips?

**Comment by jay** [comment]
> We are running into this issue with PHP specifically

AFAICT PHP [doesn't have an option](https://www.php.net/manual/en/function.curl-setopt.php) that maps to CURLOPT_SSL_CTX_FUNCTION, which means you would have to modify the PHP source (refer to #9024), modify curl source or modify your OpenSSL configuration.
